src/groupelements.jl

```diff
@@ -0,0 +1,143 @@
+    B = cocycle(a, b, inv(b))
+    return c == a * b ? B : zero(B)
+end
+
+struct ElementTable end
+"""
+    const Element
+
+A constant of a singleton type used as `Element[G,ω]` with `G<:Group` a type of group, to
+construct or obtain a concrete subtype of `AbstractElement{G,ω}` that implements the data
+structure used to represent elements of the group `G` with associator given by the
+3-cocycle `ω`.
+"""
+const Element = ElementTable()
+
+type_repr(::Type{<:AbstractGroupElement{G, ω}}) where {G <: Group, ω} = "Element[" * type_repr(G) * ", " * sprint(show, ω) * "]"
+function Base.show(io::IO, c::AbstractGroupElement)
+    I = typeof(c)
+    return if get(io, :typeinfo, nothing) !== I
+        print(io, type_repr(I), "(")
+        for k in 1:fieldcount(I)
+            k > 1 && print(io, ", ")
+            print(io, getfield(c, k))
+        end
+        print(io, ")")
+    else
+        fieldcount(I) > 1 && print(io, "(")
+        for k in 1:fieldcount(I)
+            k > 1 && print(io, ", ")
+            print(io, getfield(c, k))
+        end
+        fieldcount(I) > 1 && print(io, ")")
+    end
+end
+
+
+# ZNElement: elements of Z_N are labelled by integers mod N; do we ever want N > 64?
+"""
+    struct ZNElement{N, ω} <: AbstractGroupElement{ℤ{N}, ω}
+    ZNElement{N, ω}(n::Integer)
+    Element[ℤ{N}, ω](n::Integer)
+
+Represents an element of the group ``ℤ_N`` for some value of `N<64`. (We need 2*(N-1) <= 127 in
+order for a ⊗ b to work correctly.) For `N` equals `2`, `3` or `4`, `ℤ{N}` can be replaced
+by `ℤ₂`, `ℤ₃`, `ℤ₄`. An arbitrary `Integer` `n` can be provided to the constructor, but only
+the value `mod(n, N)` is relevant. The cocycle `ω` should also be specified as an integer 
+` 0 <= p < N`, and then leads to the 3-cocycle being given by 
+
+```julia
+cocycle(a, b, c) = cis(2* π* p * a.n * (b.n + c.n - mod(b.n + c.n, N)) / N))
+```
+
+## Fields
+- `n::Int8`: the integer label of the element, modulo `N`.
+"""
+struct ZNElement{N, p} <: AbstractGroupElement{ℤ{N}, p}
+    n::Int8
+    function ZNElement{N, p}(n::Integer) where {N, p}
+        @assert N < 64
+        @assert 0 <= p < N
+        return new{N, p}(mod(n, N))
+    end
+end
+ZNElement{N}(n::Integer) where {N} = ZNElement{N, 0}(n)
+Base.getindex(::ElementTable, ::Type{ℤ{N}}, p::Int) where {N} = ZNElement{N, mod(p, N)}
+Base.convert(::Type{ZNElement{N, p}}, n::Real) where {N, p} = ZNElement{N, p}(n)
+const Z2Element{p} = ZNElement{2, p}
+const Z3Element{p} = ZNElement{3, p}
+const Z4Element{p} = ZNElement{4, p}
+
+Base.one(::Type{ZNElement{N, p}}) where {N, p} = ZNElement{N, p}(0)
+Base.inv(c::ZNElement{N, p}) where {N, p} = ZNElement{N, p}(-c.n)
+Base.:*(c1::ZNElement{N, p}, c2::ZNElement{N, p}) where {N, p} =
+    ZNElement{N, p}(mod(c1.n + c2.n, N))
+
+function cocycle(a::ZNElement{N, p}, b::ZNElement{N, p}, c::ZNElement{N, p}) where {N, p}
+    return cis(2 * π * p * a.n * (b.n + c.n - mod(b.n + c.n, N)) / N)
```

Comment:
```suggestion
    return cispi(2 * p * a.n * (b.n + c.n - mod(b.n + c.n, N)) / N)
```

Reply:
I'll add this in together with the other changes.